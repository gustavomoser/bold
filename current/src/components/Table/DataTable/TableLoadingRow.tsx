import React, { CSSProperties } from 'react'

import { Theme, useStyles } from '../../../styles'
import { Spinner } from '../../Spinner'
import { TableCell, TableRow } from '../Table'

export interface TableLoadingRowProps {
  colSpan: number
  message?: string
}

export function TableLoadingRow(props: TableLoadingRowProps) {
  const { colSpan, message } = props
  const { classes } = useStyles(createStyles)

  return (
    <TableRow>
      <TableCell colSpan={colSpan} style={classes.cell}>
        <div className={classes.container}>
          <Spinner />
          {message}
        </div>
      </TableCell>
    </TableRow>
  )
}

TableLoadingRow.defaultProps = {
  message: 'Carregando resultados',
} as Partial<TableLoadingRowProps>

const createStyles = (theme: Theme) => ({
  cell: {
    background: theme.pallete.surface.background,
    color: theme.pallete.primary.main,
    fontWeight: 'bold',
  } as CSSProperties,
  container: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    '& > *:first-of-type': {
      marginRight: '0.5rem',
    },
  } as CSSProperties,
})
