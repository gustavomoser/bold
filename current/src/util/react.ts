import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
