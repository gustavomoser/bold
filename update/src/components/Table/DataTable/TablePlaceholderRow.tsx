import React, { CSSProperties } from 'react'

import { Theme, useStyles } from '../../../styles'
import { TableCell, TableRow } from '../Table'

export interface TablePlaceholderRowProps {
  colSpan: number
  message?: string
}

export function TablePlaceholderRow(props: TablePlaceholderRowProps) {
  const { colSpan, message } = props
  const { classes } = useStyles(createStyles)

  return (
    <TableRow>
      <TableCell colSpan={colSpan} style={classes.cell}>
        {message}
      </TableCell>
    </TableRow>
  )
}

TablePlaceholderRow.defaultProps = {
  message: 'Nenhum registro encontrado',
} as Partial<TablePlaceholderRowProps>

export const createStyles = (theme: Theme) => ({
  cell: {
    color: theme.pallete.text.secondary,
    fontStyle: 'italic',
    textAlign: 'center',
  } as CSSProperties,
})
