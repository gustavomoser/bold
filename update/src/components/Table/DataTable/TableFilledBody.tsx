import React from 'react'

import { useCss } from '../../../styles'
import { TableBody, TableCell, TableRow } from '../Table'

import { DataTableProps, defaultColumnStyles } from './DataTable'
import { TableLoadingRow } from './TableLoadingRow'
import { TablePlaceholderRow } from './TablePlaceholderRow'

export interface TableFilledBodyProps extends Pick<DataTableProps, 'columns' | 'rows' | 'loading' | 'onRowClick'> {}

export function TableFilledBody(props: TableFilledBodyProps) {
  const { columns, rows, loading, onRowClick } = props
  const { css } = useCss()

  const handleClick = row => e => onRowClick(row)

  const isEmpty = () => !rows || rows.length === 0

  return (
    <TableBody>
      {loading && <TableLoadingRow colSpan={columns.length} />}

      {!loading && isEmpty() && <TablePlaceholderRow colSpan={columns.length} />}

      {rows.map((row, idx) => (
        <TableRow key={idx} onClick={onRowClick && handleClick(row)}>
          {columns.map((col, colIdx) => (
            <TableCell key={colIdx} style={css(defaultColumnStyles(col), col.style)}>
              {col.render(row)}
            </TableCell>
          ))}
        </TableRow>
      ))}
    </TableBody>
  )
}
