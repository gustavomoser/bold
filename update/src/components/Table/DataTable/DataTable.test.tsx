import { fireEvent, render } from '@testing-library/react'
import React from 'react'

import { DataTable, DataTableProps } from './DataTable'
import { TableLoadingRow } from './TableLoadingRow'

interface Row {
  id: number
  name: string
  age: number
}

const rows: Row[] = [
  { id: 1, name: 'MARIA MACHADO DE JESUS', age: 42 },
  { id: 2, name: 'JOSÃ‰ DA SILVA MOREIRA', age: 34 },
  { id: 3, name: 'ALICE BARBOSA', age: 27 },
]

const sortHandler = jest.fn()
const DataTableTest = (props: Partial<DataTableProps>) => (
  <DataTable
    rows={rows}
    onSortChange={sortHandler}
    sort={['id', '-name']}
    columns={[
      { name: 'id', header: 'ID', sortable: true, render: (row: Row) => row.id },
      { name: 'name', header: 'Name', sortable: true, render: (row: Row) => row.name },
      {
        name: 'age',
        header: 'Age',
        sortable: true,
        align: 'right',
        style: { color: 'red' },
        render: (row: Row) => row.age,
      },
    ]}
    {...props}
  />
)

it('should render correctly', () => {
  const { container } = render(<DataTableTest />)
  expect(container).toMatchSnapshot()
})

it('should call onSortChange with right parameters when clicked over column title', () => {
  const { container } = render(<DataTableTest />)
  fireEvent.click(container.querySelector('th[data-name="id"] span'))
  expect(sortHandler).toHaveBeenLastCalledWith(['-id'])

  fireEvent.click(container.querySelector('th[data-name="name"] span'))
  expect(sortHandler).toHaveBeenLastCalledWith(['name'])

  fireEvent.click(container.querySelector('th[data-name="id"] span'), { shiftKey: true })
  expect(sortHandler).toHaveBeenLastCalledWith(['-id', '-name'])

  fireEvent.click(container.querySelector('th[data-name="age"] span'), { shiftKey: true })
  expect(sortHandler).toHaveBeenLastCalledWith(['id', '-name', 'age'])
})

it('should render TableLoadingRow when loading prop is on', () => {
  const queryText = TableLoadingRow.defaultProps.message
  expect(render(<DataTable rows={[]} columns={[]} loading={false} />).queryAllByText(queryText).length).toEqual(0)
  expect(render(<DataTable rows={[]} columns={[]} loading={true} />).queryAllByText(queryText).length).toEqual(1)
})

it('should accept the render prop', () => {
  const renderSpy = jest.fn()
  render(<DataTableTest render={renderSpy} />)
  expect(renderSpy).toHaveBeenCalledWith(
    expect.objectContaining({
      columns: expect.anything(),
    })
  )
})

describe('onRowClick prop', () => {
  it('should call the prop with the clicked row', () => {
    const clickHandler = jest.fn()
    const { container } = render(<DataTableTest onRowClick={clickHandler} />)
    const trs = container.querySelectorAll('tbody tr')

    expect(clickHandler).not.toHaveBeenCalled()

    fireEvent.click(trs[0])
    expect(clickHandler).toHaveBeenLastCalledWith(rows[0])

    fireEvent.click(trs[trs.length - 1])
    expect(clickHandler).toHaveBeenLastCalledWith(rows[rows.length - 1])
  })
})

describe('getColumn', () => {
  const renderSpy = jest.fn()
  render(<DataTableTest render={renderSpy} />)
  const getColumn = renderSpy.mock.calls[0][0].getColumn

  it('should return the column configuration by its name', () => {
    expect(getColumn('id').name).toEqual('id')
    expect(getColumn('id').sortable).toEqual(true)
  })

  it('should return undefined for inexistent columns', () => {
    expect(getColumn('foo')).toBeUndefined()
  })
})

describe('getHeaderProps', () => {
  const renderSpy = jest.fn()
  render(<DataTableTest render={renderSpy} />)
  const getHeaderProps = renderSpy.mock.calls[0][0].getHeaderProps

  it('should return the table header props of a column configuration', () => {
    const config = getHeaderProps({
      name: 'id',
      render: () => null,
      sortable: true,
      align: 'right',
    })

    expect(config).toMatchObject({
      key: 'id',
      'data-name': 'id',
    })
    expect(config).toHaveProperty('onSortChange')
  })

  it('should throw an error for inexistent columns', () => {
    expect(() => getHeaderProps('foo')).toThrowError()
  })
})
