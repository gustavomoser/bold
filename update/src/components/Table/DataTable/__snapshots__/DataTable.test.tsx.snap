// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.emotion-0 {
  width: 100%;
  border-collapse: collapse;
  border: 1px solid #D3D4DD;
}

.emotion-0 tr:hover td {
  background: #F0F0F5;
}

.emotion-3 {
  text-align: left;
  padding: 0.5rem 1rem;
  border-bottom: 1px solid #D3D4DD;
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.emotion-5 {
  margin-left: 0.25rem;
  font-size: 1rem;
  fill: #AAAAB9;
}

.emotion-6 {
  fill: #0069D0;
}

.emotion-13 {
  text-align: left;
  padding: 0.5rem 1rem;
  border-bottom: 1px solid #D3D4DD;
  text-align: right;
  color: red;
}

.emotion-20 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #D3D4DD;
}

.emotion-22 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #D3D4DD;
  text-align: right;
  color: red;
}

<div>
  <table
    class="emotion-0"
  >
    <thead
      class="emotion-1"
    >
      <tr
        class="emotion-1"
      >
        <th
          class="emotion-3"
          data-name="id"
        >
          <span
            class="emotion-4"
          >
            ID
            <svg
              class="emotion-5"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
            >
              <polygon
                class="emotion-6"
                points="12 4.94 16.95 9.89 18.36 8.47 12 2.11 5.64 8.47 7.05 9.89 12 4.94"
              />
              <polygon
                class="emotion-1"
                points="12 19.06 7.05 14.11 5.64 15.53 12 21.89 18.36 15.53 16.95 14.11 12 19.06"
              />
            </svg>
          </span>
        </th>
        <th
          class="emotion-3"
          data-name="name"
        >
          <span
            class="emotion-4"
          >
            Name
            <svg
              class="emotion-5"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
            >
              <polygon
                class="emotion-1"
                points="12 4.94 16.95 9.89 18.36 8.47 12 2.11 5.64 8.47 7.05 9.89 12 4.94"
              />
              <polygon
                class="emotion-6"
                points="12 19.06 7.05 14.11 5.64 15.53 12 21.89 18.36 15.53 16.95 14.11 12 19.06"
              />
            </svg>
          </span>
        </th>
        <th
          class="emotion-13"
          data-name="age"
        >
          <span
            class="emotion-4"
          >
            Age
            <svg
              class="emotion-5"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
            >
              <polygon
                class="emotion-1"
                points="12 4.94 16.95 9.89 18.36 8.47 12 2.11 5.64 8.47 7.05 9.89 12 4.94"
              />
              <polygon
                class="emotion-1"
                points="12 19.06 7.05 14.11 5.64 15.53 12 21.89 18.36 15.53 16.95 14.11 12 19.06"
              />
            </svg>
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      class="emotion-1"
    >
      <tr
        class="emotion-1"
      >
        <td
          class="emotion-20"
        >
          1
        </td>
        <td
          class="emotion-20"
        >
          MARIA MACHADO DE JESUS
        </td>
        <td
          class="emotion-22"
        >
          42
        </td>
      </tr>
      <tr
        class="emotion-1"
      >
        <td
          class="emotion-20"
        >
          2
        </td>
        <td
          class="emotion-20"
        >
          JOSÃ‰ DA SILVA MOREIRA
        </td>
        <td
          class="emotion-22"
        >
          34
        </td>
      </tr>
      <tr
        class="emotion-1"
      >
        <td
          class="emotion-20"
        >
          3
        </td>
        <td
          class="emotion-20"
        >
          ALICE BARBOSA
        </td>
        <td
          class="emotion-22"
        >
          27
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
